x-pack/plugin/security/qa/jwt-realm/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRestIT.java

```diff
@@ -0,0 +1,458 @@
+            deleteRoleMapping(roleMapping);
+        }
+    }
+
+    public void testFailureOnExpiredJwt() throws Exception {
+        final String principal = randomPrincipal();
+        { // Test with valid time
+            final SignedJWT jwt = buildAndSignJwtForRealm1(principal, Instant.now());
+            final TestSecurityClient client = getSecurityClient(jwt, Optional.empty());
+            assertThat(client.authenticate(), hasEntry(User.Fields.USERNAME.getPreferredName(), principal));
+        }
+
+        { // Test with expired time
+            final SignedJWT jwt = buildAndSignJwtForRealm1(principal, Instant.now().minus(12, ChronoUnit.HOURS));
+            TestSecurityClient client = getSecurityClient(jwt, Optional.empty());
+
+            // This fails because the JWT is expired
+            final ResponseException exception = expectThrows(ResponseException.class, client::authenticate);
+            assertThat(exception.getResponse(), hasStatusCode(RestStatus.UNAUTHORIZED));
+        }
+    }
+
+    public void testFailureOnInvalidRsaSignature() throws Exception {
+        final String originalPrincipal = randomPrincipal();
+        final SignedJWT originalJwt = buildAndSignJwtForRealm1(originalPrincipal, Instant.now());
+        {
+            // Test with valid signed JWT
+            final TestSecurityClient client = getSecurityClient(originalJwt, Optional.empty());
+            assertThat(client.authenticate(), hasEntry(User.Fields.USERNAME.getPreferredName(), originalPrincipal));
+        }
+
+        final String alternatePrincipal = "not-" + originalPrincipal;
+        final SignedJWT alternateJwt = buildAndSignJwtForRealm1(alternatePrincipal, Instant.now());
+        {
+            // Test with an alternate, valid signed JWT
+            final TestSecurityClient client = getSecurityClient(alternateJwt, Optional.empty());
+            assertThat(client.authenticate(), hasEntry(User.Fields.USERNAME.getPreferredName(), alternatePrincipal));
+        }
+
+        // Test that we get a failure if we combine the payload from JWT1 with the signature from JWT2
+        {
+            assertThat(
+                getSecurityClient(alternateJwt, Optional.empty()).authenticate(),
+                hasEntry(User.Fields.USERNAME.getPreferredName(), alternatePrincipal)
+            );
+
+            final Base64URL[] originalParts = SignedJWT.split(originalJwt.serialize());
+            assertThat(originalParts, arrayWithSize(3));
+
+            final Base64URL[] alternateParts = SignedJWT.split(alternateJwt.serialize());
+            assertThat(alternateParts, arrayWithSize(3));
+
+            assertThat("JWT Signatures should be different", originalParts[2], not(equalTo(alternateParts[2])));
+
+            // Create a new JWT with the header + payload from the original JWT, but a signature from the alternate JWT
+            final SignedJWT falsifiedJwt = new SignedJWT(originalParts[0], originalParts[1], alternateParts[2]);
```

Comment:
Nit. Would it be faster to generate an invalid signaure via `new Base64URL(randomAlphaOfLength(32))` instead of using alternateParts[2]?

I am wondering why sign and test alternateJwt. If the intention is to create a different signature, then perhaps a random signature is sufficient, and it will run faster. If the intention is to check some other condition, then I am curious what is it.

If I were to guess, are you testing against a specific crypto weakness pattern? Or is this type of test recommended on a best practice list? I am just looking for clarification, so I can make sure I cover the same thing elsewhere.

Reply:
I changed this to mimic the existing unit test that falsifies the claims without updating the signature.